sklearn/utils/extmath.py

```diff
@@ -839,18 +745,29 @@ def _incremental_mean_and_var(X, last_mean, last_variance, last_sample_count):
     # new = the current increment
     # updated = the aggregated stats
     last_sum = last_mean * last_sample_count
-    new_sum = _safe_accumulator_op(np.nansum, X, axis=0)
+    if sample_weight is not None:
+        new_sum = _safe_accumulator_op(np.nansum, X * sample_weight[:, None],
+                                       axis=0)
+        new_sample_count = np.sum(sample_weight[:, None] * (~np.isnan(X)),
+                                  axis=0)
+    else:
+        new_sum = _safe_accumulator_op(np.nansum, X, axis=0)
+        new_sample_count = np.sum(~np.isnan(X), axis=0)
 
-    new_sample_count = np.sum(~np.isnan(X), axis=0)
     updated_sample_count = last_sample_count + new_sample_count
 
     updated_mean = (last_sum + new_sum) / updated_sample_count
 
     if last_variance is None:
         updated_variance = None
     else:
-        new_unnormalized_variance = (
-            _safe_accumulator_op(np.nanvar, X, axis=0) * new_sample_count)
+        if sample_weight is not None:
+            T = new_sum / new_sample_count
+            new_unnormalized_variance = np.nansum(sample_weight[:, None] *
+                                                  (X - T)**2, axis=0)
+        else:
+            new_unnormalized_variance = (
+                _safe_accumulator_op(np.nanvar, X, axis=0) * new_sample_count)
```

Comment:
This is potentially the cause of the slowdown for `sample_weight=None` case:

```python
>>> %timeit np.nansum((X - m) ** 2, axis=0)
420 ms ± 9.36 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
>>> %timeit np.nanvar(X, axis=0)
528 ms ± 3.27 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

100 ms difference on the full X. We might want to always use `np.nansum` in both branches and report the suboptimal performance of `np.nanvar` to the numpy issue tracker.

Reply:
Note that the non-nan variants of those functions have opposite relative performance profiles:

```
>>> %timeit np.sum((X - m) ** 2, axis=0)
257 ms ± 1.54 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
>>> %timeit np.var(X, axis=0)
247 ms ± 1.02 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Reply:
Is there a reason not to do the following ?
```suggestion
        T = new_sum / new_sample_count
        if sample_weight is not None:
            new_unnormalized_variance = np.nansum(sample_weight[:, None] *
                                                  (X - T)**2, axis=0)
        else:
            new_unnormalized_variance = np.nansum((X - T)**2, axis=0)
```

Reply:
@jeremiedbb do you mean to compute the variance by hand? If this is the case I would rather use it `np.nanvar` and add a comment that `np.nanvar` does not support `sample_weight` and thus we rely to compute by hand?

Reply:
I ran @ogrisel 's script (10 times) in my environment (Intel(R) Xeon(R) CPU E5-2690 v4 @ 2.60GHz).
 
||Main branch|This PR|
|---|---|---|
|`scale_with_partial_fit(sample_weight=None)`|1.975517841684632|1.9935310563072561|
|`scale_with_partial_fit(sample_wieght)`|2.9529636044986547|2.231434507598169|

In my environment, the PR version is much faster.

Reply:
@glemaitre well we don't compute the var, but ``var * n_samples`` which is ``sum((X-X_mean)**2)``.

Reply:
Thanks @nuka137 for the bench results. It's great to see that this PR is always faster. Can you please try to benchmark @jeremiedbb's suggestion above to see if it can improve the performance for the `sample_weight is None` case?

Reply:
> @jeremiedbb do you mean to compute the variance by hand? If this is the case I would rather use it np.nanvar.

@glemaitre the point is to see if we can improve the performance by not calling `np.nanvar` as it recomputes the mean which we have already computed.

Reply:
Uhm I see the point of @ogrisel. Fine with calling the `nansum`

Reply:
@ogrisel 

Yeah, @jeremiedbb 's suggestion is much faster.

||Main branch|This PR|This PR (+@jeremiedbb's suggestion)|
|---|---|---|---|
|`scale_with_partial_fit(sample_weight=None)`|1.975517841684632|1.9935310563072561|**1.7161642136983573**|
|`scale_with_partial_fit(sample_wieght)`|2.9529636044986547|2.231434507598169|2.268341443617828|

Reply:
@nuka137 thanks for the benchmark! +1 for updating this PR accordingly.

Reply:
@ogrisel Thanks, I committed @jeremiedbb 's suggestion code.

Reply:
Interestingly, on my machine I get consistently better times for `sample_weight=None`, but in the case of presence of `sample_weight`, for me main is faster than PR and updated PR:
```
main
%timeit scale_with_partial_fit(X, sample_weight=sample_weight)
1.26 s ± 21 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

PR pre update
%timeit scale_with_partial_fit(X, sample_weight=sample_weight)
1.4 s ± 15.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

PR post update
%timeit scale_with_partial_fit(X, sample_weight=sample_weight)
1.4 s ± 14.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
No idea why ...

Reply:
It looks to me that there are quite some statistical fluctuations here ;-)

Reply:
@ogrisel can we consider this review part as resolved?